x-pack/plugins/enterprise_search/server/routes/app_search/engines.ts

```diff
@@ -9,38 +9,52 @@ import { schema } from '@kbn/config-schema';
 
 import { ENGINES_PAGE_SIZE } from '../../../common/constants';
 
-export function registerEnginesRoute({ router, config }) {
+export function registerEnginesRoute({ router, config, log }) {
   router.get(
     {
       path: '/api/app_search/engines',
       validate: {
         query: schema.object({
-          type: schema.string(),
+          type: schema.oneOf([schema.literal('indexed'), schema.literal('meta')]),
           pageIndex: schema.number(),
         }),
       },
     },
     async (context, request, response) => {
-      const appSearchUrl = config.host;
-      const { type, pageIndex } = request.query;
-
-      const url = `${appSearchUrl}/as/engines/collection?type=${type}&page[current]=${pageIndex}&page[size]=${ENGINES_PAGE_SIZE}`;
-      const enginesResponse = await fetch(url, {
-        headers: { Authorization: request.headers.authorization },
-      });
-
-      if (enginesResponse.url.endsWith('/login')) {
-        return response.ok({
-          body: { message: 'no-as-account' },
-          headers: { 'content-type': 'application/json' },
+      try {
+        const appSearchUrl = config.host;
+        const { type, pageIndex } = request.query;
+
+        const url = `${appSearchUrl}/as/engines/collection?type=${type}&page[current]=${pageIndex}&page[size]=${ENGINES_PAGE_SIZE}`;
+        const enginesResponse = await fetch(url, {
+          headers: { Authorization: request.headers.authorization },
         });
-      }
 
-      const engines = await enginesResponse.json();
-      return response.ok({
-        body: engines,
-        headers: { 'content-type': 'application/json' },
-      });
+        if (enginesResponse.url.endsWith('/login')) {
+          log.info('No corresponding App Search account found');
+          // Note: Can't use response.unauthorized, Kibana will auto-log out the user
+          return response.forbidden({ body: 'no-as-account' });
+        }
+
+        const engines = await enginesResponse.json();
+        const hasValidData =
+          Array.isArray(engines?.results) && typeof engines?.meta?.page?.total_results === 'number';
+
+        if (hasValidData) {
+          return response.ok({
+            body: engines,
+            headers: { 'content-type': 'application/json' },
+          });
+        } else {
+          // Either a completely incorrect Enterprise Search host URL was configured, or App Search is returning bad data
+          throw new Error(`Invalid data received from App Search: ${JSON.stringify(engines)}`);
+        }
+      } catch (e) {
+        log.error(`Cannot connect to App Search: ${e.toString()}`);
+        if (e instanceof Error) log.debug(e.stack);
+
+        return response.notFound({ body: 'cannot-connect' });
```

Comment:
Should we distinguish here between `cannot-connect` and some other malformed response error condition?

Reply:
We can - I can't think of anything super useful to show the user in that case though 🤔

I think the assumption that malformed data from App Search = 'This isn't App Search' is a relatively safe one in terms of end-user-experience, as we can't show them anything but an error message in that case anyway. Definitely open to other thoughts or suggestions however!

Reply:
Good point -- not much we can show the user. The more adventurous user or support engineer can still glean valuable lower-level detail by poking into the stack trace that we're logging to the console. 👍